text/stacks/0007-squash-stack.md

```diff
@@ -0,0 +1,45 @@
+# Squash Stack Images
+
+## Summary
+
+The build and run images delivered as a stack for our users should be squashed
+into a single layer to reduce the total number of layers in the builder and
+application images and remove duplicated files that appear in multiple layers.
+
+## Motivation
+
+Stack images act as the base image for applications and our own builder images.
+Their internal layer implementations should not be important to end users.
+Recently, we have been fielding a lot of bug reports wherein users cannot
+create their application images because the total number of layers for the
+image has exceeded the limit enforced by `containerd`.
+
+Squashing the stack images down into a single layer will create more headroom
+for both the builders and the images they create.
+
+## Detailed Explanation
+
+The `jam create-stack` command will include a new flag, `--squash`, that will
+squash the build and run image layers into a single layer. These squashed
+images will be what the project ships as the canonical stack images for Bionic
+and Jammy.
+
+## Rationale and Alternatives
```

Comment:
Instead of `jam create-stack --squash` we could modify the pipelines to run `crane flatten`.

This has this advantage of not requiring us to maintain more code in `jam create-stack`.

Reply:
I think the main issue here is that you need to run this against an image that is already in a registry. So, we'd need to push to some registry, flatten, pull, and then push everywhere else? And then secondarily, nobody else in the community would really benefit from this. Is that fine?

Reply:
Oh, ok. I didn't dig into the mechanism of `crane flatten`. We could explore pushing the image to a local registry, flattening, then pushing to dockerhub/gcr. But I agree that this path is sounding less attractive than adding support in `jam create-stack`.